### PR TITLE
WSTEAM1-994 - Adds custom file upload html and state handling

### PR DESCRIPTION
**Pull request by HarveyPeachey**
Resolves JIRA [WSTEAM1-994](https://jira.dev.bbc.co.uk/browse/WSTEAM1-994)

Overall changes
======
Adds a custom file upload display, which allows management of adding and removing multiple files adding previews for images and icons for other file types.

Code changes
======

- Adds a custom file uploader button that invokes the hidden file input
- Changes state storage for files, to store them as an array of Files, instead of the input value
- Adjusts submission logic for files for state storage change
- Adds an ID for field labels, so `aria-describedby` can target labels
- Adds SVG icons for Upload, Delete
- Adds SVG icons as dataURI strings for Video, Audio and Documents
- Adds adds state and logic for removal of files
- Adds state and file reader logic for displaying thumbnails for images
- Adds styling for file component
- Adds unit tests

Testing
======
1. cd into `ws-nextjs-app`
2. Run `yarn`
3. Run `yarn dev`
4. Visit http://localhost:7081/somali/send/u130092370?renderer_env=test

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx [hunk 1]
```diff
@@ -1,26 +1,152 @@
...
+          }),
+          ...(!isInitialUpload && {
+            value: [...(prevState[name].value as File[]), ...filesArray],
+          }),
+        },
+      };
+
+      return { ...prevState, ...updatedState };
+    });
+  };
```
Comment (amoore108) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Took a bit of inspiration from https://github.com/KunalN25/multiple-file-uploads/blob/main/src/App.js and came up with this to make this function a bit simpler:
```
  const { isValid, required } = inputState;
  const { setFormState } = useFormContext();
  const inputRef = useRef<HTMLInputElement>(null);
  const filesInState = inputState.value as File[];

  const handleFileChange = (e: ChangeEvent<HTMLInputElement>) => {
    const chosenFiles = Array.prototype.slice.call(e.target.files) as File[];
    const uploaded = [...filesInState];

    chosenFiles.some(file => {
      if (uploaded.findIndex(f => f.name === file.name) === -1) {
        uploaded.push(file);
      }
      return null;
    });

    setFormState(prevState => {
      return {
        ...prevState,
        [name]: { ...prevState[name], value: uploaded },
      };
    });
  };
```

Seems to work when I try it, but I'm not sure if it covers all the edge cases!

Comment (amoore108) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: In fact you may just be able to use `handleChange` thats passed in the component args and not use the `setFormState` function directly anymore:

```
  const handleFileChange = (e: ChangeEvent<HTMLInputElement>) => {
    const chosenFiles = Array.prototype.slice.call(e.target.files) as File[];
    const uploaded = [...filesInState];

    chosenFiles.some(file => {
      if (uploaded.findIndex(f => f.name === file.name) === -1) {
        uploaded.push(file);
      }
      return null;
    });

    handleChange(name, uploaded);
  };
```

Comment (amoore108) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: I haven't tried actually submitting these files, so this all may be non-applicable 😅

Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Yeah was on the fence with `Array.prototype.slice.call` partly because I thought being explicit would help understand why can't just manipulate the FileList, happy to change to it

Comment (amoore108) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Yea I needed to read up on it myself. I think overall its maybe a bit simpler to reason about the function, but up to other reviewers as well if the prototype function is a bit too magic-y.

Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: I've opted to change it, and just added a comment above it

Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Hmm just thought about that some method, there might be a case where a user is uploading a file with the same name, but are different? And I don't think we have an error state for that, as the old uploader allowed uploading of multiple of the same file

Comment (amoore108) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Does the `file.name` include the extension as well? If it does, I think it shouldn't allow that scenario, don't think you can have 2 files named the same with the same extension as the OS will usually add in `(1)` or something to the end of the name to differentiate them.

Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: You can if it lives in a different directory, as the file object contains a reference to the file location on a users machine. So if bob wanted to upload `/downloads/img.jpg`, then opens the uploader again and adds `/desktop/img.jpg` it would let them do it

Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: So i guess we can use that value

Comment (amoore108) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File.tsx]: Ah yea right enough. Tbh I think its fine to let them upload multiple of the same file. Probably an edge case that we don't need to worry too much about and would add complexity to cater for.

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        />,
+      );
+    });
+
+    expect(
+      screen.getByText(/here's what you're sending:/i),
+    ).toBeInTheDocument();
+  });
+
+  it('should call the function to update state when a file is added', async () => {
```
Comment (HarveyPeachey) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx:112]: This test is a little brittle, as it's not testing anything in the UI, due to state management being handled in the the context, only the arguments within the `handleChange` function are being tested here

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+    const inputFile = container.querySelector('#foo') as Element;
+    fireEvent.change(inputFile, { target: { files: [mockImgFile] } });
+
+    expect(contextValue.handleChange).toHaveBeenCalledWith('bar', [
+      mockImgFile,
+    ]);
+  });
+
+  it('should remove a file from the list when the remove button is clicked', async () => {
```
Comment (HarveyPeachey) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx:135]: Same as previous brittle comment

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      );
+    });
+
+    const uploadButton = screen.getByRole('button', {
+      name: /choose a file/i,
+    });
+    await user.click(uploadButton);
+
+    expect(inputRefClick).toHaveBeenCalled();
+  });
```
Comment (HarveyPeachey) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.test.tsx]: Not sure what I'm doing wrong here, think I've stared at tests too much. Apparently the click on the input ref isn't being called, maybe I've mocked it incorrectly with Spy?

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/styles.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      display: 'none',
+    }),
+  fileUploadButton: ({ spacings, palette, fontVariants, fontSizes }: Theme) => {
+    const topBottomPadding = `${spacings.HALF + spacings.FULL}rem`;
+    const leftRightPadding = `${spacings.DOUBLE}rem`;
+
+    return css({
+      padding: `${topBottomPadding} ${leftRightPadding} ${topBottomPadding} ${leftRightPadding}`,
+      border: 'none',
+      backgroundColor: '#0071F1',
```
Comment (amoore108) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File/styles.ts]: Think we can use the new Theme colours for this and the hover/focus state:
```
export const SERVICE_NEUTRAL_CORE = '#0071F1';
export const SERVICE_NEUTRAL_DARK = '#0051AD';
```

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        >
+          <DeleteSvg />
+          <VisuallyHiddenText>Remove</VisuallyHiddenText>
+        </button>
+      </li>
+    );
+  });
+  return (
+    <>
+      <p css={styles.fileListParagraph}>
```
Comment (amoore108) [inline ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.tsx]: Could maybe use the Paragraph component here:
```
<Paragraph fontVariant='' size=''>
```

and remove the custom styling as the font styling will be set by the component.

**Comment by amoore108** [comment]
Looks really great. Huge piece of a work here, so great job tackling this 👏 

Is it worth adding a Storybook story for this component too? Think its maybe the only input type that doesn't have a story.

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/styles.ts (hunk 5 above)
Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File/styles.ts]: Good spot, done.

File: ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.tsx (hunk 6 above)
Comment (HarveyPeachey) [reply ws-nextjs-app/pages/[service]/send/[id]/FormField/File/index.tsx]: Done

**Comment by HarveyPeachey** [comment]
> Looks really great. Huge piece of a work here, so great job tackling this 👏
> 
> Is it worth adding a Storybook story for this component too? Think its maybe the only input type that doesn't have a story.

Thank ya, I've added a story and the data for the somali page I've been testing on

**Review by amoore108 — APPROVED** [review]

**Review by shayneahchoon — APPROVED** [review]
